### DISCUSSION TITLE
Best way to open Snacks.dashboard when all buffers are closed

### DISCUSSION THREAD

**Discussion by DaniBedz** [folke/snacks.nvim#279] [Q&A]
Just wondering if anyone has a good way to automatically open the dashboard when all named buffers are closed?

I tried to do this by an auto-command, but it throws invalid buffer id errors the second time I go to the dashboard and open a new buffer, as well as hiding my lualine.

Here's what I tried:
```
vim.api.nvim_create_augroup("DashboardAutoOpen", { clear = true })
vim.api.nvim_create_autocmd({ "BufDelete" }, {
	group = "DashboardAutoOpen",
	callback = function()
		vim.schedule(function()
			local buffers = vim.tbl_filter(function(buf)
				return vim.api.nvim_buf_is_valid(buf) and vim.bo[buf].buflisted and vim.api.nvim_buf_get_name(buf) ~= ""
			end, vim.api.nvim_list_bufs())

			if #buffers == 0 then
				vim.cmd("lua Snacks.dashboard()")
			end
		end)
	end,
})
```
Any guidance would be very much appreciated.

**Comment by SuXYIO** [comment]
Your autocmd works pretty good for me, i didn't see any errors, except that a empty buffer is created after opening a file via your autocmd opened dashboard

**Reply by DaniBedz** [reply]
Thanks for taking a look.

That's very strange, I'm able to recreate the errors with this very minimal repro (just lazy, Snacks and the autocmd):

https://github.com/DaniBedz/nvim-minimal

All I do is:

1. Open Neovim (dashboard shows automatically)
2. Open a buffer
3. Close it (dashboard is toggled on by the autocmd)
4. Open another buffer 

```
Error executing vim.schedule lua callback: ...ni/.local/share/nvim/lazy/snacks.nvim/lua/snacks/win.lua:499: Invalid buffer id: 4
stack traceback:
[C]: in function 'nvim_win_set_buf'
...ni/.local/share/nvim/lazy/snacks.nvim/lua/snacks/win.lua:499: in function <...ni/.local/share/nvim/lazy/snacks.nvim/lua/snacks/win.lua: 498>
```

**Comment by dikesh** [comment]
I am also getting the same error.

![image](https://github.com/user-attachments/assets/debd37bc-00c3-408a-a30e-d15ca4bec5b6)

**Comment by SuXYIO** [comment]
okay, seems like i can reproduce it too. also the terminal section in dashboard is still showing up, lualine is still closed (probably has something to do with my lualine set to hide on filetype `snacks_dashboard`), nvimtree can't open too, with similar error.  
what is more interesting, i have to quit twice to quit nvim in the new buffer. 
![screenshot](https://github.com/user-attachments/assets/7765d274-e342-4155-b71e-f705747b9197)

**Comment by SuXYIO** [comment]
i think your buffer count code has no big problems, since just running `lua Snacks.dashboard()` leads to similar problems. `lua Snacks.dashboard.open()` doesn't help either.  
So i experimented with window numbers. the initial window number was 1000, and was 1013 when i opened the dashboard via `:lua Snacks.dashboard()`, and was still 1013 when i ran `:e foo.lua`. That explains the double quit.  
Maybe we should close the dashboard first before opening a new file. But i didn't think up of any elegant way, creating another autocmd seems redundant. I'm now trying to mimic the behavior of a normal dashboard file open https://github.com/folke/snacks.nvim/blob/972c61cc1cd254ef3b43ec1dfd51eefbdc441a7d/lua/snacks/dashboard.lua#L851
```lua
vim.api.nvim_create_autocmd({ "BufRead" }, {
	group = "DashboardAutoOpen",
	callback = function()
		if Snacks.dashboard.status.opened == true then
			-- mimic normal dashboard fileopen
		end
	end
})
```
im really not an expert on this :sweat_smile:

**Reply by DaniBedz** [reply]
Creating another autocmd seems janky and throws additional errors for me. Did you have an example that works?

I'd love if this was a built-in config option, I would have thought that this is a pretty common use case.

I set this up with `dashboard-nvim` fine - but I really wanted to use Snacks.dashboard instead really.

Happy to donate to anyone that can get this working.
